src/compiler/checker.ts

```diff
@@ -6805,8 +6808,10 @@ namespace ts {
             return !!getPropertyOfType(type, "0");
         }
 
-        function isStringLiteralType(type: Type) {
-            return type.flags & TypeFlags.StringLiteral;
+        function isStringLiteralUnionType(type: Type): boolean {

```

Comment:
It can be either a single string literal type or a union of string literal types. I'm not sure your suggestion better covers that.